### ISSUE TITLE
meta not indicated as reserved crate name when shadowing module

### ISSUE THREAD

**Issue by yodaldevoid** [rust-lang/rust#73948]
Taken from this internals thread: https://internals.rust-lang.org/t/is-the-module-name-meta-forbidden/9587

A module named `meta` is shadowed leading to misleading error messages. `meta` was reserved as a built-in crate name in 2018 edition, but the error message does not indicate this.

I tried this code:

```rust
mod meta {
    pub struct Thing;
}

use meta::Thing;

fn main() {}
```
Above code compiled with `rustc --edition 2018 {file}`

This is expected to indicated that `meta` is a reserved crate name.

Instead, the following error is reported:
```
error[E0463]: can't find crate for `meta`
 --> testy.rs:5:5
  |
5 | use meta::Thing;
  |     ^^^^ can't find crate

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
```

### Meta
`rustc --version --verbose`:
```
rustc 1.44.1 (c7087fe00 2020-06-17)
binary: rustc
commit-hash: c7087fe00d2ba919df1d813c040a5d47e43b0fe7
commit-date: 2020-06-17
host: x86_64-unknown-linux-gnu
release: 1.44.1
LLVM version: 9.0
```

This issue has been assigned to @chocol4te via [this comment](https://github.com/rust-lang/rust/issues/73948#issuecomment-654795936).

**Comment by chocol4te** [comment]
@rust-highfive claim

**Comment by eddyb** [comment]
It's not forbidden as a module name, this obviously works:
```rust
mod meta {
    pub struct Thing;
}

use crate::meta::Thing;
```

I think the confusion arises from that fact that this works without `crate::`:
```rust
mod foo {
    pub struct Thing;
}

use foo::Thing;
```

But those are uniform imports (or at least I think that's what they're called), and only resolve to the surrounding scope if unambiguous (which it is for `meta` due to the.

I'm guessing the right fix here is to add more information to the error suggesting that if the intention wasn't to refer to a crate, `crate::` or `self::` etc. should/can be used explicitly.

cc @petrochenkov

**Comment by yodaldevoid** [comment]
@eddyb 

I believe the `crate::` "fix" was mentioned in a reddit thread linking to the internals thread, but I forgot to mention it. I will reword the issue as it is misworded, as you pointed out. Lastly, what you suggested for a fix lines up with what I was expecting.

Also, now you know how deep my backlog of todo items is.
